### ISSUE TITLE
research - re-using Js class globally (or speeding up init time)

### ISSUE THREAD

**Issue by tomholub** [FlowCrypt/flowcrypt-android#166]
We should first measure it:
 - how long does it take to init (try in a loop and measure)
 - how long does it take to perform a cheap operation (parse a public key - measure in a loop)

**Comment by DenBond7** [comment]
Some of the scripts take a lot of time (v8.executeScript();). But I don't think that we can speed up that very much.

D/JS: read window.js to RAM = 1 ms
D/JS: executeScript: window.js = 5 ms

_**D/JS: read openpgp.js to RAM = 34 ms
D/JS: executeScript: openpgp.js = 506 ms**_

D/JS: read punycode.js to RAM = 1 ms
D/JS: executeScript: punycode.js = 6 ms

_**D/JS: read emailjs-stringencoding.js to RAM = 31 ms
D/JS: executeScript: emailjs-stringencoding.js = 197 ms**_

D/JS: read emailjs-addressparser.js to RAM = 1 ms
D/JS: executeScript: emailjs-addressparser.js = 5 ms

D/JS: read emailjs-mime-codec.js to RAM = 3 ms
D/JS: executeScript: emailjs-mime-codec.js = 11 ms

D/JS: read emailjs-mime-parser.js to RAM = 2 ms
D/JS: executeScript: emailjs-mime-parser.js = 7 ms

D/JS: read emailjs-mime-types.js to RAM = 5 ms
D/JS: executeScript: emailjs-mime-types.js = 25 ms

D/JS: read emailjs-mime-builder.js to RAM = 2 ms
D/JS: executeScript: emailjs-mime-builder.js = 7 ms

D/JS: read mnemonic.js to RAM = 1 ms
D/JS: executeScript: mnemonic.js = 6 ms

D/JS: read global.js to RAM = 0 ms
D/JS: executeScript: global.js = 1 ms

_**D/JS: read common.js to RAM = 8 ms
D/JS: executeScript: common.js = 58 ms**_

I created JsForUiManager. It is configuring when we start the app. While the process of the app is alive Js tool is available for the UI thread. That solution gives more speed.

**Comment by DenBond7** [comment]
![image](https://user-images.githubusercontent.com/2863246/33991524-9b6577ae-e0d7-11e7-816c-b0a9e7d2661b.png)

**Comment by DenBond7** [comment]
```
D/JS: V8.createV8Runtime() = 230 ms
D/JS: bindJavaMethods() = 1 ms
D/JS: loadJavascriptCode() = 924 ms
D/JS: bindCallbackCatcher() = 0 ms
```

**Comment by DenBond7** [comment]
#161

**Comment by tomholub** [comment]
Wow, that's really, really slow. This will be a great improvement.

**Comment by DenBond7** [comment]
![image](https://user-images.githubusercontent.com/2863246/34052634-e29fe3ee-e1cb-11e7-9ad1-5668baa95bf3.png)

A text decryption is not fast too. I used a simple MIME message with the text "This is a very security encrypted text." Therefore, for decryption, I need a separate thread.

**Comment by DenBond7** [comment]
Link to [`JsTest`](https://github.com/FlowCrypt/flowcrypt-android/blob/968ecdc3f0b40e119eb1aee448be8307bff90c9b/FlowCrypt/src/androidTest/java/com/flowcrypt/email/js/JsTest.java)

**Comment by tomholub** [comment]
The slow part is the public key decryption. This is done for every message regardless of size. The symmetric encryption part is quite fast. So you won't see much difference between a tiny message and a very large message

**Comment by tomholub** [comment]
I'm closing this as a lot of the ideas were implemented in 0.6.0

We can create more issues if there is anything left unresolved.
